block-producer/src/block_builder/prover/block_witness.rs

```diff
@@ -246,6 +188,147 @@ impl BlockWitness {
+        // from the bottom to the top
+        let mut stack_inputs = Vec::new();
+
+        // Chain MMR stack inputs
+        {
+            stack_inputs.extend(self.prev_header.hash());
+            stack_inputs.extend(self.chain_peaks.hash_peaks());
+        }
+
+        // Nullifiers stack inputs
+        {
+            let num_produced_nullifiers: u64 = self
+                .produced_nullifiers
+                .len()
+                .try_into()
+                .expect("can't be more than 2^64 - 1 nullifiers");
+
+            for nullifier in self.produced_nullifiers.keys() {
+                stack_inputs.extend(*nullifier);
+            }
+
+            // append nullifier value (`[0, 0, 0, block_num]`)
+            let block_num = self.prev_header.block_num() + 1;
+            stack_inputs.extend([ZERO, ZERO, ZERO, block_num.into()]);
+
+            // append initial nullifier root
+            stack_inputs.extend(self.prev_header.nullifier_root());
+
+            // append number of nullifiers
+            stack_inputs.push(num_produced_nullifiers.into());
+        }
+
+        // Notes stack inputs
+        {
+            let num_created_notes_roots = self.batch_created_notes_roots.len();
+            for (batch_index, batch_created_notes_root) in self.batch_created_notes_roots.iter() {
+                stack_inputs.extend(batch_created_notes_root.iter());
+
+                let batch_index =
+                    u64::try_from(*batch_index).expect("can't be more than 2^64 - 1 notes created");
+                stack_inputs.push(Felt::from(batch_index));
+            }
+
+            let empty_root = EmptySubtreeRoots::entry(CREATED_NOTES_TREE_DEPTH, 0);
+            stack_inputs.extend(*empty_root);
+            stack_inputs.push(Felt::from(
+                u64::try_from(num_created_notes_roots)
+                    .expect("can't be more than 2^64 - 1 notes created"),
+            ));
+        }
+
+        // Account stack inputs
+        let mut num_accounts_updated: u64 = 0;
+        for (idx, (&account_id, account_update)) in self.updated_accounts.iter().enumerate() {
+            stack_inputs.push(account_id.into());
+            stack_inputs.extend(account_update.final_state_hash);
+
+            let idx = u64::try_from(idx).expect("can't be more than 2^64 - 1 accounts");
+            num_accounts_updated = idx + 1;
+        }
+
+        // append initial account root
+        stack_inputs.extend(self.prev_header.account_root());
+
+        // append number of accounts updated
+        stack_inputs.push(num_accounts_updated.into());
+
+        StackInputs::new(stack_inputs)
+    }
+
+    /// Builds the advice inputs to the block kernel
+    fn build_advice_inputs(self) -> Result<AdviceInputs, BlockProverError> {
+        let advice_map: Vec<_> = self
+            .produced_nullifiers
+            .values()
+            .map(|proof| (proof.leaf().hash().as_bytes(), proof.leaf().to_elements()))
+            .collect();
```

Comment:
Adding nullifier-related info the the advice inputs happens here and also down starting at line 317. I wonder if we should encapsulate it better. Maybe do something similar to what we do with the chain MMR? Specifically, we could have a separate small helper function - something like `add_nullifiers_to_advice_inputs()` and put all relevant logic there.

Reply:
went with a slightly different approach. `add_nullifiers_to_advice_inputs()` would not have worked as well because it needs to modify the `MerkleStore`, and `AdviceInputs::extend_merkle_store()` takes an iterator of `InnerNodeInfo` (which would be a pain to generate). I also didn't like that we would build the `AdviceInputs`, and then mutate them once again (I would even remove that functionality from `AdviceInputs`'s API - but that's another discussion).

This new approach instead combines the building of the merkle store, and the building of the advice map; as opposed to grouping nullifiers and accounts together, respectively. I find that even better actually - when investigating problems with the store (or advice map), you only have to look in one place.